core/src/main/scala/magnolia1/magnolia.scala

```diff
@@ -640,29 +639,50 @@ object Magnolia {
             .fold(error(_), identity)
         }
 
-        val assignments = typeclasses.zipWithIndex.map { case ((subType, typeclass), idx) =>
-          val symbol = subType.typeSymbol
-          val (annotationTrees, inheritedAnnotationTrees) = annotationsOf(symbol)
-          q"""$subtypesVal($idx) = $SubtypeObj[$typeConstructor, $genericType, $subType](
-              ${typeNameOf(subType)},
-              $idx,
-              $ArrayObj(..${annotationTrees}): Array[_root_.scala.Any],
-              $ArrayObj(..${inheritedAnnotationTrees}): Array[_root_.scala.Any],
-              $ArrayObj(..${typeAnnotationsOf(symbol, fromParents = true)}): Array[_root_.scala.Any],
-              $CallByNeedObj($typeclass),
-              (t: $genericType) => t.isInstanceOf[$subType],
-              (t: $genericType) => t.asInstanceOf[$subType]
-            )"""
-        }
-
         val subType = appliedType(SubtypeTpe, typeConstructor, genericType)
+        val groupSize = 500
+        val partialSubtypesVal = c.freshName(TermName("partialSubtypes"))
+        val (functionNames, partialAssignmentFunctions) = typeclasses
+          .grouped(groupSize)
+          .toList
+          .map { group =>
+            val name = c.freshName(TermName("partialAssignments"))
+            val assignments = group.zipWithIndex.map { case ((subType, typeclass), idx) =>
+              val symbol = subType.typeSymbol
+              val (annotationTrees, inheritedAnnotationTrees) = annotationsOf(symbol)
+              q"""$partialSubtypesVal($idx) = $SubtypeObj[$typeConstructor, $genericType, $subType](
+                        ${typeNameOf(subType)},
+                        $idx,
+                        $ArrayObj(..${annotationTrees}): Array[_root_.scala.Any],
+                        $ArrayObj(..${inheritedAnnotationTrees}): Array[_root_.scala.Any],
+                        $ArrayObj(..${typeAnnotationsOf(symbol, fromParents = true)}): Array[_root_.scala.Any],
+                        $CallByNeedObj($typeclass),
+                        (t: $genericType) => t.isInstanceOf[$subType],
+                        (t: $genericType) => t.asInstanceOf[$subType]
+                      )"""
+            }
+            val tree = q"""def $name = {
+                val $partialSubtypesVal = new $ArrayClass[$subType](${assignments.size})
+                ..$assignments
+                $partialSubtypesVal
+              }
+           """
+
+            (name, tree)
+          }
+          .unzip
+
+        val subtypesVal = c.freshName(TermName("subtypes"))
+        val combinations = functionNames.map(name => q"""$subtypesVal ++= $name""")
+
         Some(q"""{
-          val $subtypesVal = new $ArrayClass[$subType](${assignments.size})
-          ..$assignments
+          ..$partialAssignmentFunctions
+          val $subtypesVal = Array.newBuilder[$subType]
```

Comment:
Pass the array and index to the helper methods, no need for a builder

Reply:
[4acf714](https://github.com/softwaremill/magnolia/pull/485/commits/4acf7140732b01d41f7ec61c9ad2bc639c260704)